### PR TITLE
Feature tam control

### PR DESCRIPTION
**Pull request by coamitch**
## Changes Made

Please provide a description of all changes made in this PR and why the changes
are needed.

## Associated Issues

Please provide a list of all open issues that this PR will close or contribute
toward closing.

- Fixes # (issue)

## Testing

Please provide a clear and concise description of the testing performed.

### CODE REVIEW

File: thruster_allocation_matrix_controller/src/thruster_allocation_matrix_controller.cpp [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+  const rclcpp_lifecycle::State & /*previous_state*/)
+{
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+controller_interface::return_type ThrusterAllocationMatrixController::update_and_write_commands(
+  const rclcpp::Time & /*time*/, const rclcpp::Duration & /*period*/)
+{
+  // getting the reference forces and calculating the thrust forces using the tam
+  const std::vector<double> reference_forces_values(reference_interfaces_.begin() + dof_, reference_interfaces_.end());
```
Comment (evan-palmer) [inline thruster_allocation_matrix_controller/src/thruster_allocation_matrix_controller.cpp:137]: Why are you indexing `+ dof_`? you only have up to `dof_` in the vector.

File: thruster_allocation_matrix_controller/src/thruster_allocation_matrix_controller.cpp [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+  auto reference_forces = create_six_dof_eigen_from_named_vector(dof_names_, six_dof_names_, reference_forces_values);
+
+  Eigen::VectorXd thruster_forces = tam_.completeOrthogonalDecomposition().pseudoInverse() * reference_forces;
+
+  // setting command interfaces with calculated thruster forces
+  for (std::int64_t i = 0; i < num_thrusters_; i++) {
+    command_interfaces_[i].set_value(thruster_forces[i]);
+  }
+
+  // TODO: we still aren't sure if this is the correct type/way of doing this. we may need to come back and do
```
Comment (evan-palmer) [inline thruster_allocation_matrix_controller/src/thruster_allocation_matrix_controller.cpp:147]: Add this back in before you submit the PR

File: thruster_allocation_matrix_controller/src/thruster_allocation_matrix_controller.cpp [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+    (tam_y.size() != num_thrusters_) || (tam_z.size() != num_thrusters_) || (tam_rx.size() != num_thrusters_) ||
+    (tam_ry.size() != num_thrusters_) || (tam_rz.size() != num_thrusters_)) {
+    // if any of the tam vectors aren't the same size, then we need to notify the user and return failure
+    RCLCPP_ERROR(get_node()->get_logger(), "Received vectors for TAM with differing lengths.");
+    return controller_interface::CallbackReturn::ERROR;
+  }
+
+  // otherwise, they are all the same length and we can pack them into an Eigen matrix
+  Eigen::MatrixXd tam_(dof_, num_thrusters_);
+  tam_ << tam_x, tam_y, tam_z, tam_rx, tam_ry, tam_rz;
```
Comment (evan-palmer) [inline thruster_allocation_matrix_controller/src/thruster_allocation_matrix_controller.cpp:238]: Check this in your debugging. I think Eigen is column major so things might be inserted weird.

File: thruster_allocation_matrix_controller/CMakeLists.txt [hunk 4]
```diff
@@ -1,26 +1,87 @@
...
 find_package(ament_cmake REQUIRED)
-# uncomment the following section in order to fill in
-# further dependencies manually.
-# find_package(<dependency> REQUIRED)
+foreach(Dependency IN ITEMS ${THIS_PACKAGE_INCLUDE_DEPENDS})
+  find_package(${Dependency} REQUIRED)
+endforeach()
+
+generate_parameter_library(thruster_allocation_matrix_parameters
+  src/thruster_allocation_matrix_parameters.yaml
```
[the remark this answers is not in the document]
Comment (rakeshv24) [reply thruster_allocation_matrix_controller/CMakeLists.txt:28]: Fixed

File: thruster_allocation_matrix_controller/include/thruster_allocation_matrix_controller/thruster_allocation_matrix_controller.hpp [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+  controller_interface::CallbackReturn on_activate(const rclcpp_lifecycle::State & previous_state) override;
+
+  THRUSTER_ALLOCATION_MATRIX_CONTROLLER_PUBLIC
+  controller_interface::CallbackReturn on_deactivate(const rclcpp_lifecycle::State & previous_state) override;
+
+  THRUSTER_ALLOCATION_MATRIX_CONTROLLER_PUBLIC
+  controller_interface::return_type update_and_write_commands(
+    const rclcpp::Time & time, const rclcpp::Duration & period) override;
+
+  // TODO(evan): Might not need to override this
```
[the remark this answers is not in the document]
Comment (rakeshv24) [reply thruster_allocation_matrix_controller/include/thruster_allocation_matrix_controller/thruster_allocation_matrix_controller.hpp:88]: Deleted.

File: thruster_allocation_matrix_controller/include/thruster_allocation_matrix_controller/thruster_allocation_matrix_controller.hpp [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+  size_t dof_;
+
+  // rows for tam matrix
+  Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic> tam_;
+
+  std::int64_t num_thrusters_;
+
+private:
+  THRUSTER_ALLOCATION_MATRIX_CONTROLLER_LOCAL
+  // TODO: replace message type with something for the Wrench
```
[the remark this answers is not in the document]
Comment (rakeshv24) [reply thruster_allocation_matrix_controller/include/thruster_allocation_matrix_controller/thruster_allocation_matrix_controller.hpp:136]: Deleted.

File: thruster_allocation_matrix_controller/include/thruster_allocation_matrix_controller/thruster_allocation_matrix_controller.hpp [hunk 7]
```diff
@@ -0,0 +1,140 @@
+// Copyright 2024, Evan Palmer
```
[the remark this answers is not in the document]
Comment (rakeshv24) [reply thruster_allocation_matrix_controller/include/thruster_allocation_matrix_controller/thruster_allocation_matrix_controller.hpp:1]: Noted

File: thruster_allocation_matrix_controller/package.xml [hunk 8]
```diff
@@ -8,9 +8,27 @@
   <license>MIT</license>
 
   <buildtool_depend>ament_cmake</buildtool_depend>
+  <buildtool_depend>eigen3_cmake_module</buildtool_depend>
```
[the remark this answers is not in the document]
Comment (rakeshv24) [reply thruster_allocation_matrix_controller/package.xml:11]: Not required and deleted.

File: thruster_allocation_matrix_controller/src/thruster_allocation_matrix_controller.cpp [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+#include "thruster_allocation_matrix_controller/thruster_allocation_matrix_controller.hpp"
+
+#include <Eigen/Dense>
+
+#include "hardware_interface/types/hardware_interface_type_values.hpp"
+
+namespace thruster_allocation_matrix_controller
+{
+
+namespace
```
[the remark this answers is not in the document]
Comment (rakeshv24) [reply thruster_allocation_matrix_controller/src/thruster_allocation_matrix_controller.cpp:30]: Agreed but we can work on this later.

File: thruster_allocation_matrix_controller/src/thruster_allocation_matrix_controller.cpp [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+  }
+
+  // Initialize the reference and state realtime messages
+  const std::shared_ptr<geometry_msgs::msg::Wrench> reference_msg = std::make_shared<geometry_msgs::msg::Wrench>();
+  reference_.writeFromNonRT(reference_msg);
+
+  // Subscribe to the reference topic
+  reference_sub_ = get_node()->create_subscription<geometry_msgs::msg::Wrench>(
+    "~/reference", rclcpp::SystemDefaultsQoS(),
+    [this](const std::shared_ptr<geometry_msgs::msg::Wrench> msg) { reference_callback(msg); });  // NOLINT
```
[the remark this answers is not in the document]
Comment (rakeshv24) [reply thruster_allocation_matrix_controller/src/thruster_allocation_matrix_controller.cpp:108]: Not entirely sure what you mean. Are you suggesting not to have a subscriber callback to set the value?

File: thruster_allocation_matrix_controller/test/test_load_thruster_allocation_matrix_controller.cpp [hunk 11]
```diff
@@ -0,0 +1,60 @@
+// Copyright (c) 2021, PickNik, Inc.
```
[the remark this answers is not in the document]
Comment (rakeshv24) [reply thruster_allocation_matrix_controller/test/test_load_thruster_allocation_matrix_controller.cpp:1]: Fixed it.

File: thruster_allocation_matrix_controller/test/test_thruster_allocation_matrix_controller.cpp [hunk 12]
```diff
@@ -0,0 +1,297 @@
+// // Copyright (c) 2021, Stogl Robotics Consulting UG (haftungsbeschränkt)
```
[the remark this answers is not in the document]
Comment (rakeshv24) [reply thruster_allocation_matrix_controller/test/test_thruster_allocation_matrix_controller.cpp:1]: Fixed it.

File: thruster_allocation_matrix_controller/test/test_load_thruster_allocation_matrix_controller.cpp
[the remark this answers is not in the document]
Comment (rakeshv24) [reply thruster_allocation_matrix_controller/test/test_load_thruster_allocation_matrix_controller.cpp:1]: done

File: thruster_allocation_matrix_controller/test/test_thruster_allocation_matrix_controller.cpp
[the remark this answers is not in the document]
Comment (rakeshv24) [reply thruster_allocation_matrix_controller/test/test_thruster_allocation_matrix_controller.cpp:1]: done

File: thruster_allocation_matrix_controller/test/test_thruster_allocation_matrix_controller.hpp
[the remark this answers is not in the document]
Comment (rakeshv24) [reply thruster_allocation_matrix_controller/test/test_thruster_allocation_matrix_controller.hpp:1]: done

File: thruster_allocation_matrix_controller/src/thruster_allocation_matrix_controller.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (evan-palmer) [reply thruster_allocation_matrix_controller/src/thruster_allocation_matrix_controller.cpp:108]: The lambda function is already the callback. What you have here is:

subscribe to -> callback (lambda) which calls -> reference_callback

If reference_callback isn't too involved then you can just put the contents of reference_callback in the lambda.

File: thruster_allocation_matrix_controller/thruster_allocation_matrix_controller.xml [hunk 13]
```diff
@@ -0,0 +1,11 @@
+<library path="thruster_allocation_matrix_controller">
+
+  <class name="thruster_allocation_matrix_controller/ThrusterAllocationMatrixController"
+    type="thruster_allocation_matrix_controller::ThrusterAllocationMatrixController"
+    base_class_type="controller_interface::ChainableControllerInterface">
+    <description>
+      Thruster allocation matrix controller for wrench to thrust control.
```
[the remark this answers is not in the document]
Comment (rakeshv24) [reply thruster_allocation_matrix_controller/thruster_allocation_matrix_controller.xml:7]: fixed

File: thruster_allocation_matrix_controller/src/thruster_allocation_matrix_controller.cpp [hunk 14]
```diff
@@ -0,0 +1,259 @@
...
+  }
+
+  // otherwise, they are all the same length and we can pack them into an Eigen matrix
+  Eigen::MatrixXd tam_(dof_, num_thrusters_);
+  tam_ << tam_x, tam_y, tam_z, tam_rx, tam_ry, tam_rz;
+
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+void ThrusterAllocationMatrixController::reference_callback(std::shared_ptr<geometry_msgs::msg::Wrench> msg)
```
[the remark this answers is not in the document]
Comment (rakeshv24) [reply thruster_allocation_matrix_controller/src/thruster_allocation_matrix_controller.cpp:243]: Not sure. Same comment as the other relevant comment.

File: thruster_allocation_matrix_controller/src/thruster_allocation_matrix_controller.cpp [hunk 15]
```diff
@@ -0,0 +1,259 @@
...
+      vec[i] = values[std::distance(dof_names.begin(), it)];
+    }
+  }
+
+  return vec;
+}
+
+}  // namespace
+
+ThrusterAllocationMatrixController::CallbackReturn ThrusterAllocationMatrixController::on_init()
```
[the remark this answers is not in the document]
Comment (rakeshv24) [reply thruster_allocation_matrix_controller/src/thruster_allocation_matrix_controller.cpp:58]: Not sure what you mean.

File: thruster_allocation_matrix_controller/src/thruster_allocation_matrix_controller.cpp (hunk 1 above)
Comment (rakeshv24) [reply thruster_allocation_matrix_controller/src/thruster_allocation_matrix_controller.cpp:137]: Fixed.

File: thruster_allocation_matrix_controller/src/thruster_allocation_matrix_controller.cpp (hunk 2 above)
Comment (rakeshv24) [reply thruster_allocation_matrix_controller/src/thruster_allocation_matrix_controller.cpp:147]: Noted.

File: thruster_allocation_matrix_controller/src/thruster_allocation_matrix_controller.cpp (hunk 3 above)
Comment (rakeshv24) [reply thruster_allocation_matrix_controller/src/thruster_allocation_matrix_controller.cpp:238]: Okay, will do.

File: .vscode/settings.json [hunk 16]
```diff
@@ -6,7 +6,83 @@
     "*.srdf": "xml",
     "*.rviz": "yaml",
     "*.config": "xml",
-    "*.sdf": "xml"
+    "*.sdf": "xml",
```
[the remark this answers is not in the document]
Comment (rakeshv24) [reply .vscode/settings.json]: Outdated because of the rebase?

**Pull request closed without merging**